src/tss2-fapi/ifapi_ima_eventlog.c

```diff
@@ -0,0 +1,853 @@
+
+/** Callback for digest with name of used hash algorithm,
+ */
+static TSS2_RC
+digest_with_hash_name_cb(UINT8 *digest, UINT8 *buffer, size_t *offset, json_object *jso,
+                        IFAPI_IMA_TEMPLATE *template) {
+    TSS2_RC r;
+    char hash_alg[CRYPTO_MAX_ALG_NAME + 1] = { 0 };
+    size_t alg_name_size;
+    const EVP_MD *md;
+    int digest_size;
+    UINT32 digest_buffer_size;
+
+    digest_buffer_size = get_size_from_buffer(buffer, offset);
+    alg_name_size = strlen((char *)&buffer[*offset]) - 1; /**< strip : */
+    if (alg_name_size > CRYPTO_MAX_ALG_NAME) {
+        return_error(TSS2_FAPI_RC_BAD_VALUE, "Invalid hash name.");
+    }
+    memcpy(hash_alg, &buffer[*offset], alg_name_size);
+    md = EVP_get_digestbyname(hash_alg);
+    if (md == NULL) {
+        return_error(TSS2_FAPI_RC_BAD_VALUE, "Invalid hash name.");
+    }
+    *offset += alg_name_size + 2; /**< skip : and '\0' */
+    digest_size = EVP_MD_size(md);
+    if (alg_name_size + 2 + digest_size != digest_buffer_size) {
+        return_error(TSS2_FAPI_RC_BAD_VALUE, "Invalid IMA binary format.");
+    }
+
+    LOGBLOB_TRACE(&buffer[*offset], digest_size, "IMA data_hash");
+
+    if (jso && zero_digest(digest, template->hash_size) &&
+        zero_digest(&buffer[*offset], digest_size)) {
+        r = set_ff_digest(jso);
+        return_if_error(r, "Set 0xff in digest.");
+    }
+    *offset += digest_size;
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to get digest with size field (UINT32).
+ */
+static TSS2_RC
+signature_cb(UINT8 *digest, UINT8 *buffer, size_t *offset, json_object *jso,
+             IFAPI_IMA_TEMPLATE *template) {
+    (void) digest;
+    (void) jso;
+    UINT32 digest_size;
+    (void)template;
+
+    digest_size =  get_size_from_buffer(buffer, offset);
+    LOGBLOB_TRACE(&buffer[*offset], digest_size, "IMA Signature:");
+    *offset += digest_size;
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to get null terminated name with size field (n-ng).
+ */
+static TSS2_RC eventname_ng_json_cb(UINT8 *digest, UINT8 * buffer, size_t *offset, json_object *jso,
+                                    IFAPI_IMA_TEMPLATE *template) {
+    size_t size;
+    UINT32 size_from_buffer;
+    (void)digest;
+    (void)jso;
+
+    /* Get size from buffer without 0 Terminator. */
+    size_from_buffer = get_size_from_buffer(buffer, offset) - 1;
+    size = strlen((const char *)&buffer[*offset]);
+    if (size != size_from_buffer) {
+        return_error2(TSS2_FAPI_RC_BAD_VALUE,
+                      "Invalid digest size, string length: %zu size from buffer: %"
+                      PRIu32, size, size_from_buffer);
+    }
+    LOG_TRACE("IMA name: %s", (const char *)&buffer[*offset]);
+    template->name = (char *)&buffer[*offset];
+    *offset += size + 1; /**< with 0 terminator */
+    return TSS2_RC_SUCCESS;
+}
+
+/** Callback to get null terminated name (n).
+ */
+static TSS2_RC eventname_cb(UINT8 *digest, UINT8 * buffer, size_t *offset, json_object *jso,
```

Comment:
I just noticed this now, inconsistent spacing on `*`. Use it tight to the variable everywhere. `UINT8 *buffer`. Here and throughout any others I missed. This seems not explicitly covered in the coding style doc, but its by and large the convention in this code base.